// @skimhugo/eslint-config-basic@0.2.8/import.cjs
// Inline from https://github.com/standard/eslint-config-standard/blob/master/.eslintrc.json
// Until it upgrades

module.exports = {
  parserOptions: {
    ecmaVersion: 2022,
    ecmaFeatures: {
      jsx: true,
    },
    sourceType: 'module',
  },
  env: {
    es2021: true,
    node: true,
  },
  plugins: [
    'import',
  ],
  globals: {
    document: 'readonly',
    navigator: 'readonly',
    window: 'readonly',
  },
  rules: {
// analysis/correctness
    'import/no-unresolved': 'off',
    'import/named': 'error',
    'import/namespace': 'error',
    'import/default': 'error',
    'import/export': 'error',
    'import/no-duplicates': 'error',

    // red flags (thus, warnings)
    'import/no-named-as-default': 'warn',
    'import/no-named-as-default-member': 'warn',


    // import
    'import/first': 'error',
    'import/no-absolute-path': ['error', { esmodule: true, commonjs: true, amd: false }],
    'import/no-named-default': 'error',
    'import/no-webpack-loader-syntax': 'error',

    // import
    'import/order': 'error',
    // 'import/first': 'error',
    'import/no-mutable-exports': 'error',
    // 'import/no-absolute-path': 'off',

    // 'import/no-named-as-default-member': 'off',
    // 'import/no-named-as-default': 'off',
    // 'import/namespace': 'off',

    /**
     * 禁止重复导入模块
     */
    'no-duplicate-imports': 'error',
  },
}
