// @skimhugo/eslint-config-basic@0.2.8/promise.cjs
// Inline from https://github.com/standard/eslint-config-standard/blob/master/.eslintrc.json
// Until it upgrades

module.exports = {
  parserOptions: {
    ecmaVersion: 2022,
    ecmaFeatures: {
      jsx: true,
    },
    sourceType: 'module',
  },
  env: {
    es2021: true,
    node: true,
  },
  plugins: [
    'promise',
  ],
  globals: {
    document: 'readonly',
    navigator: 'readonly',
    window: 'readonly',
  },
  rules: {

    'promise/param-names': 'error',
    /**
     * 禁止在 return 语句里使用 await
     */
    'no-return-await': 'error',
    /**
     * async 函数中必须存在 await 语句
     */
    'require-await': 'off',
    /**
     * 禁止将 async 函数做为 new Promise 的回调函数
     * @reason 出现这种情况时，一般不需要使用 new Promise 实现异步了
     */
    'no-async-promise-executor': 'error',
    // sequence promise need temporarily open this eslint
    /**
     * 禁止将 await 写在循环里，因为这样就无法同时发送多个异步请求了
     * @reason 要求太严格了，有时需要在循环中写 await
     */
    'no-await-in-loop': 'error',
    /**
     * 禁止在 Promise 的回调函数中直接 return
     */
    'no-promise-executor-return': 'error',
    // make sure the variable is read at the same time it's updated.
    /**
     * 禁止将 await 或 yield 的结果做为运算符的后面项
     * @reason 这样会导致不符合预期的结果
     * https://github.com/eslint/eslint/issues/11899
     * 在上面 issue 修复之前，关闭此规则
     */
    'require-atomic-updates': 'error',
    'max-nested-callbacks': ['error', 3],
    /**
     * Promise 的 reject 中必须传入 Error 对象，而不是字面量
     */
    'prefer-promise-reject-errors': 'error',

  },
}
